alr init: add compiler option controls to template project

Fabien-Chouteau opened the pull request:
User can control compile checks, run-time checks, contracts, style checks.

Fabien-Chouteau committed:
alr init: add compiler option controls to template project

User can control compile checks, run-time checks, contracts, style
checks.

mosteo commented on src/alr/alr-commands-init.adb:
```diff
@@ -79,30 +81,84 @@ package body Alr.Commands.Init is
+         Put_Line ("                            ""-gnatn""); -- Enable inlining");
+         Put_Line ("      when ""debug"" =>");
+         Put_Line ("         Build_Switches := (""-g"",   -- Debug info");
+         Put_Line ("                            ""-O0""); -- No optimization");
```
What do you think about using `-Og` instead of `-O0`? According to https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html

> It is a better choice than -O0 for producing debuggable code because some compiler passes that collect debug information are disabled at -O0.

Personally, I haven't noticed anything ever but...

Fabien-Chouteau replied:
Ok for me

mosteo commented on src/alr/alr-commands-init.adb:
```diff
@@ -63,6 +63,8 @@ package body Alr.Commands.Init is
          Filename : constant String :=
             +Full_Name (Directory / (+Lower_Name & ".gpr"));
       begin
+         --  Use more than 80 colums for more readable strings
+         pragma Style_Checks ("M200");
```
I'm going to copy this trick at some point...

Fabien-Chouteau replied:
We should avoid it :)

Fabien-Chouteau committed:
Testsuite: fix Ada coding style in tests

Fabien-Chouteau committed:
Alr.Commands.Init: use UNIX line ending on all generated source code

To pass default GNAT style checks.

Fabien-Chouteau committed:
Update alr-commands-init.adb

Alr.Commands.Init: fix line ending by using file IO binary mode

Fabien-Chouteau committed:
Alr.Commands.Init: rever line ending changes and just remove the check

Git is messing with line ending on Windows so trying to produce UNIX files is more difficult than planned.

Instead the DOS line terminator style check is disabled.

mosteo commented:
I'm out of my depth here. @Fabien-Chouteau , if a crate doesn't depend on a cross-compiler, can it still specify a runtime and/or target in its gpr file?

hgrodriguez commented:
ok I removed the dependency on the runtime and target in EEPROM_i2c.grp. please ignore this PR I will raise another one for the update. Sorry for time wasted here on your both sides.

Fabien-Chouteau commented:
Indeed the library should not specify target nor run-time, this will be set by the application that with's the library.

Fabien-Chouteau commented:
Probably something we should document under a "Portable Libraries" section...

hgrodriguez commented:
> Probably something we should document under a "Portable Libraries" section...

👍

hgrodriguez commented:
can I close the pull request myself to start over and have a fresh beginning?

mosteo commented:
I'm cloisng this one now, but yes, @hgrodriguez, you can close any PR you open.
